### PR TITLE
Allow updating photo of a contact

### LINKED ISSUE #78: As a forgetful user, I can upload photos for my contacts
So that I can visually remember them

### PR DESCRIPTION
**Pull request by VN-Hao**
This fixes [#78](https://github.com/AY2324S1-CS2103T-T13-2/tp/issues/78)

### CODE REVIEW

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.logic.parser.AddressBookParser;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.ModelManager;
+
+public class UpdatePhotoCommandTest {
+
+    @Test
+    public void invalidIndex_exceptionThrown() {
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java]: Consider renaming this to follow the [Java coding standard for test methods](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=featureUnderTest_testScenario_expectedBehavior())

Suggestion: `execute_invalidIndex_exceptionThrown`

File: src/test/java/seedu/address/logic/parser/UpdatePhotoCommandParserTest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.Command;
+import seedu.address.logic.commands.UpdatePhotoCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+public class UpdatePhotoCommandParserTest {
+
+    @Test
+    public void parseInvalidCommand_exceptionThrown() {
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/parser/UpdatePhotoCommandParserTest.java]: ditto, consider `parse_invalidCommand_exceptionThrown`

File: src/test/java/seedu/address/model/person/AvatarTest.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+public class AvatarTest {
+
+    private static final String DEFAULT_PATH = "/images/default_photo.png";
+
+    @Test
+    public void testDefaultConstructor() {
```
Comment (limjunxian1) [inline src/test/java/seedu/address/model/person/AvatarTest.java]: ditto, consider `avatar()`

File: src/main/java/seedu/address/model/person/Avatar.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public void setAvatar(String path) {
+        this.path = path;
+    }
+
+    public Image getImage() {
+        try {
+            return new Image(new FileInputStream(path));
+        } catch (FileNotFoundException e) {
```
Comment (limjunxian1) [inline src/main/java/seedu/address/model/person/Avatar.java:37]: Consider the event when the path given by the user is invalid. This results in the interface still showing a successful execution of command.

Perhaps consider having an `Image` attribute, with its default value being `new Image(this.getClass().getResourceAsStream(DEFAULT_PATH)` as per your code. So your `getImage()` method returns the `Image` itself

File: src/main/java/seedu/address/model/person/Avatar.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+public class Avatar {
+
+    private static final String DEFAULT_PATH = "/images/default_photo.png";
+    private String path;
+
+    public Avatar() {
+        this.path = DEFAULT_PATH;
+    }
+
+    public Avatar(String path) {
```
Comment (limjunxian1) [inline src/main/java/seedu/address/model/person/Avatar.java]: As per your implementation, seems like this constructor is used when creating the newly edited person after `update photo` command.

A suggestion is to change this method will be

```
public Avatar(String path) throws FileNotFoundException {
    Image image = new Image(new FileInputStream(path));
}
```

Subsequently, your private method of `copyPerson` in your `UpdatePhotoCommand` class will throw `FileNotFoundException` and will be caught by its `execute` method. In the catch block, can consider the following

```
catch (FileNotFoundException e) {
    throw new CommandException("Invalid file path provided.");
}
```

**Review by limjunxian1 — COMMENTED** [review]
Overall, the implementation of the feature LGTM. I have raised a few comments that you may wish to consider and we can further discuss on this matter. Otherwise, do try to insert more test cases to improve code coverage. As of now, your test cases are only considering failures. Perhaps, you may also want to input test cases that results in successful execution.

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        } catch (CommandException e) {
+            fail();
+        }
+    }
+
+    @Test
+    public void execute_noPathCommand_exceptionThrown() {
+        AddressBookParser parser = new AddressBookParser();
+        try {
+            Command aCommand = parser.parseCommand(
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:42]: Perhaps the naming of the command here could be better to follow the Java coding standards? e.g. `addCommand`

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                    "add n/John Doe p/98765432 e/johndoe@gmail.com a/John street, block 123, #01-01 t/friend"
+            );
+            Command upCommand = parser.parseCommand("updatephoto 1");
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:45]: ditto

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+            assertTrue(aCommand instanceof AddCommand);
+            assertTrue(upCommand instanceof UpdatePhotoCommand);
+            AddCommand addCommand = (AddCommand) aCommand;
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:48]: i see that here you are using the `addCommand` name here, but is this type casting really necessary for you to execute the command at line 51? let runtime handle which method to invoke for `execute` since polymorphism is applied

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            UpdatePhotoCommand updatePhotoCommand = (UpdatePhotoCommand) upCommand;
+            ModelManager temp = new ModelManager();
+            addCommand.execute(temp);
+            updatePhotoCommand.execute(temp);
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:52]: ditto

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        }
+    }
+
+    @Test
+    public void execute_invalidIndexInEmptyAddressBook_exceptionThrown() {
+        AddressBookParser parser = new AddressBookParser();
+        try {
+            Command command = parser.parseCommand("updatephoto 0 path/images/default_photo.png");
+            assertTrue(command instanceof UpdatePhotoCommand);
+            UpdatePhotoCommand updatePhotoCommand = (UpdatePhotoCommand) command;
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:67]: i shall not repeat my previous comments on this issue, refer to previous comments

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        } catch (ParseException e) {
+            assertEquals(e.getMessage(),
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, UpdatePhotoCommand.MESSAGE_USAGE));
+        } catch (CommandException e) {
+            fail();
+        }
+    }
+
+    @Test
+    public void execute_invalidIndexInEmptyAddressBook_exceptionThrown() {
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:62]: perhaps the idea of `InEmptyAddressBook` is not really necessary? as i look at your subsequent test cases, looks like you are trying to apply [test case boundary analysis](https://nus-cs2103-ay2324s1.github.io/website/schedule/week10/topics.html#w10-5-test-cases-boundary-value-analysis) which i believe that making an address book empty isn't really necessary. Just a suggestion

File: src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            updatePhotoCommand.execute(temp);
+        } catch (ParseException e) {
+            fail();
+        } catch (CommandException e) {
+            assertEquals(e.getMessage(), "The person index provided is invalid");
+        }
+    }
+
+    @Test
+    public void execute_invalidPath_exceptionThrown() {
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/commands/UpdatePhotoCommandTest.java:121]: Just a suggestion, testing can be afford to be more rigourous, using the idea of [equivalence partitioning](https://nus-cs2103-ay2324s1.github.io/website/schedule/week10/topics.html#w10-4-test-cases-equivalence-partitioning), e.g. consider that `invalidPath` can be an empty string as well

File: src/test/java/seedu/address/logic/parser/UpdatePhotoCommandParserTest.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.Command;
+import seedu.address.logic.commands.UpdatePhotoCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+public class UpdatePhotoCommandParserTest {
```
Comment (limjunxian1) [inline src/test/java/seedu/address/logic/parser/UpdatePhotoCommandParserTest.java:11]: Just a suggestion, can consider successful parsing of correct input as well for rigorous testing. You can follow more guidelines on what to test [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week11/topics.html#w11-3-test-cases-combining-multiple-inputs)

**Review by limjunxian1 — APPROVED** [review]
Overall, LGTM, implementation seems fine as I looked through your code. Happy to merge this even though there are some minor comments that perhaps will be beneficial for you to consider.  When this is merged, I will try to spot any potential bugs arising from your feature as well!
